models/provenance.py

```diff
@@ -259,11 +257,11 @@ def update_hash(self):
             process=self.process,
             parameters=self.parameters,
             upstream_hashes=self.upstream_hashes,
-            code_version=self.code_version.version
+            code_version=self.code_version.id
         )
         json_string = json.dumps(superdict, sort_keys=True)
 
-        self.unique_hash = base64.urlsafe_b64encode(hashlib.sha256(json_string.encode("utf-8")).digest()).decode()[:20]
+        self.id = base64.urlsafe_b64encode(hashlib.sha256(json_string.encode("utf-8")).digest()).decode()[:20]
```

Comment:
Here's something unrelated to this PR but it would be a really good opportunity to solve this:

Can we make sure this string does not contain underscores? It would make our filenames more easy to parse. 

I'm not sure what the best way to remove them. I used `urlsafe` which gets rid of other things like dashes (if I recall) and replaces them with underscores, so maybe it would be good enough just to not use `urlsafe`. 
Another option is to just decide here (the only place where we make these strings) that we will always use `replace` and change the underscores to something else. 

What do you think?

Reply:
It's sad that python's base64 encode has an altchars for + and /, but not _

One solution would be to go from base64 to md5sum.  md5sum is all hexidecimal digits.  (hexits?)

Reply:
...or, just take the b64encode off.  Replace hashlib.sha256().digest() with hashlib.sha256().hexdigest() ; same result.  20 hexidecimal digits is almost certainly going to be safe, even if it's 4x fewer combinations than 20 base64 characters.

Reply:
We *could* make the id field a UUID rather than a string.  I think that will have more efficient indexing, but indexing of the provenances is not going to be an issue.  (We're not going to have millions of provenances in the database... or, if we do, I'm very sad.)

Reply:
I was really happy to have more than 16 options per digit because I could reduce the number of digits. For example, if you shorten to 6 digits for filenames it helps to have more options. 

Do you think we could just replace the underscores with dashes or something?

Reply:
The `urlsafe` replaces + with - and / with _. I actually like having no slashes, but would prefer to replace underscore with dash. I don't see why we shouldn't just add a `replace('_', '-')` after the `urlsafe` gets rid of the dashes and slashes first.

Reply:
We could... I guess it bothers me aesthetically to have an ecoding that's not reversable (since + is already replaced with -), though the hash is already not reversable, so probably not a big deal.

I'm trying to think of any other ASCII character we could put in a filename, but all of them (other than a-zA-Z0-9, -, and _) are potentially troublesome (even though legal).  We could always just use greek letters....

How about base32? That gives you 30 bits of variation in 6 characters, which is probably enough.

Reply:
How would that work, exactly?

Reply:
Just replace base64.urlsafe_b64encode with base64.b32hexencode

Reply:
I did this, but used b32encode instead of b32hexencode ; the latter is new in Python 3.10, and the current test dockerfile has python 3.9.  (I'll be bumping that to 3.11 with one of my upcoming pull requests.)